### ISSUE TITLE
🚀 --language option as bat

### ISSUE THREAD

**Issue by jonathanpoelen** [dandavison/delta#499]
When files have no extension or come from a stream, `delta` is not able to display color

![2021-01-10-034254_651x232_scrot](https://user-images.githubusercontent.com/1436727/104113064-3801b300-52f6-11eb-965d-9d2f4758498a.png)

I suggest adding `--language` as is the case in `bat`:

![2021-01-10-034354_627x76_scrot](https://user-images.githubusercontent.com/1436727/104113067-3b953a00-52f6-11eb-9bbf-d0e6320577b3.png)

**Comment by dandavison** [comment]
Hi @jonathanpoelen, thanks for this. It's similar to https://github.com/dandavison/delta/issues/485, but that issue is focusing on the case of git diffs which can of course have multiple files and thus multiple languages. You make a good point that a `--language` flag, while not appropriate for the (most common) case of delta as a git pager, would make sense when delta is used to compare two files.

**Comment by nick87720z** [comment]
There could be way to specify language per file. It could have optional parameter for file names, probably with support for globs or regex. Multiple of them could be separated... I'm not sure, what separator char to use though - there's some example with sed 's' command; in this case first char could be used as separator, optionally limited to be only punct- or just non-alnum class to be interpreted. Default could be comma or colon.

**Comment by dandavison** [comment]
Hi @jjmaestro. `.gitattributes` support sounds interesting At the moment the best we have is `--default-language`. Note that this can be set on a per-repo basis.

https://github.com/dandavison/delta/pull/1655#issuecomment-2001999025

**Comment by jjmaestro** [comment]
> At the moment the best we have is `--default-language`. Note that this can be set on a per-repo basis: [#1655 (comment)](https://github.com/dandavison/delta/pull/1655#issuecomment-2001999025)

@dandavison thanks! I saw that in other discussions and I've tried getting it to work with `git show` without much success 😅

But yeah, even if it worked, it's a bit of a bummer that it can only be one language 😐  Would you be open to accept a PR for a `--languages` config that accepts a map of patterns that can match the filenames? I've never done anything in Rust but this is a feature I would use so I'm happy to learn and try something 😄

**Comment by dandavison** [comment]
Yes, if you'd like to work on this that would be great! Let's get the design agreed on here. cc @th1000s.

> a `--languages` config that accepts a map of patterns that can match the filenames?

Does that differ from your suggestion above to use `.gitattributes`?

If we were to go in the `.gitattributes` direction then note that delta uses the Rust `git2` bindings to `libgit2` and thus it looks to me that there may be a convenient function for us to use: https://docs.rs/git2/latest/git2/struct.Repository.html#method.get_attr

Other related docs for reference
https://libgit2.org/docs/reference/main/attr/index.html
https://git-scm.com/docs/gitattributes

**Comment by jjmaestro** [comment]
The `--languages` suggestion was just thinking about a "simpler alternative", but if you are happy to go with the `.gitattributes` solution, that's definitely better.

Re `git2`'s `get_attr`, that definitely looks exactly like what would be needed 😄  

> I'd be attracted to doing the read from disk lazily (i.e. only if the need arises).

Yeah, I don't think there would be any need to use this info in other cases except to disambiguate the case of not being able to guess the syntax. Maybe in the future someone will want something like `--always-use-gitattributes` that would override the syntax but I'd say that's better left for the future.

Then, when it comes to `.gitattributes` I think there should be a `--git-attribute` config that defaults to `delta-language` but that can be changed to e.g. `linguist-language` (or maybe even a CSV list of possible attributes in order of priority). This way, `delta` syntax can be independent of GH syntax if need be while being able to configure it to use `linguist-language`, which is an attribute that's already in use by me and [many others](https://github.com/search?q=path%3A.gitattributes%20linguist-language%3D&type=code) :) or whatever other attribute that other tools could use.

What do you think?

P.S. finally, re. "timeline", I've never worked with Rust, and I've been wanting to, so that's an incentive 😄  but I'm currently quite busy trying to finish a large project so I want to give myself some latitude 😅  I've read [CONTRIBUTING.md](https://github.com/dandavison/delta/blob/main/CONTRIBUTING.md) and [ARCHITECTURE.md](https://github.com/dandavison/delta/blob/main/ARCHITECTURE.md) (BTW, kudos for all that, I wish all projects had these :) so I'll try to play around with setting up Rust, go over the codebase and running the tests over the next week or so. I think the changes needed might be small enough that I'll be able to do them!
